Fix connection handling

feeblefakie opened the pull request:
https://scalar-labs.atlassian.net/browse/DLT-8028

Since the original code didn't really use JDBC properly, this PR fixes it.

feeblefakie commented:
@yito88 @brfrn169 It is verified by Boney for PTAL.

brfrn169 commented on src/main/java/com/scalar/cassy/server/CassyServerModule.java:
```diff
@@ -48,7 +48,6 @@ CassyServerConfig provideCassyServerConfig() {
 
   @Provides
-  @Singleton
   Connection provideConnection() throws SQLException {
```
It looks like this `provideConnection` method is no longer used as we create a `Connection` object in `CassyServerController.java` every time we need. Can we remove this method?

Similarly, the annotations `@Inject` on the constructors of `BackupHistory` and `ClusterInfo` and `RestoreHistory`  can be removed as these instances are no longer created by Guice?

feeblefakie replied:
> It looks like this provideConnection method is no longer used as we create a Connection object in CassyServerController.java every time we need. Can we remove this method?

RemoteCommandHandler still uses Connection injected by Guice, but it might be confusing to create some connections through Guice and other connections manually. What do you think?
 
> Similarly, the annotations @Inject on the constructors of BackupHistory and ClusterInfo and RestoreHistory can be removed as these instances are no longer created by Guice?

Good point! I'll remove @Inject from those. Thank you!

yito88 reviewed (approved):
LGTM!

feeblefakie commented:
@brfrn169 Sorry for the late fix.
I updated the code based on your feedback. (Also refactored it around Connection a bit)
PTAL when you get a chance 🙏

brfrn169 commented on src/main/java/com/scalar/cassy/server/RemoteCommandHandler.java:
```diff
@@ -2,33 +2,38 @@
+  private final CassyServerConfig config;
   private final BlockingQueue<RemoteCommandContext> futures;
-  private final DatabaseAccessor database;
+  private boolean isActive = true;
```
I think we should add `volatile` modifier or use `AtomicBoolean` instead of `boolean` here as this `isActive` field is accessed by multiple threads.

feeblefakie replied:
Good catch! Thank you!
I'll add a `volatile` keyword.
BTW, using `AtomicBoolean` fixes the cache issue?
(if `AtomicBoolean` is used, a written value by a thread will be guaranteed to be read by another thread immediatelyly?)

brfrn169 replied:
Yes, using `AtomicBoolean` should fix the cache issue:
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/package-summary.html

feeblefakie replied:
I didn't know that. Thank you!

brfrn169 reviewed (approved):
LGTM! Thank you!
